src/group_theory/burnside.lean

```diff
@@ -0,0 +1,153 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yury Kudryashov
+
+Prove Burnside lemma (not due to Burnside): the number of orbits of a group action equals the average number of points fixed by an element of the group.
+-/
+
+import .group_action ..order.fixed_points
+
+/- TODO: What file is a good home for this? -/
+instance lattice.fixed_points_fintype
+  {α : Type*} [fintype α] [h : decidable_eq α] (f : α → α) :
+  fintype (lattice.fixed_points f) :=
+@subtype.fintype _ _ (lattice.fixed_points f) (assume x, h (f x) x)
+
+namespace mul_action
+
+universes u v
+
+variables {G : Type u} [group G] {α : Type v} [mul_action G α]
+
+@[simp]
+lemma smul_inv_left {g : G} {x : α} : g⁻¹ • g • x = x :=
+by rw [← mul_smul, mul_left_inv, one_smul]
+
+@[simp]
+lemma smul_inv_right {g : G} {x : α} : g • g⁻¹ • x = x :=
+by rw [← mul_smul, mul_right_inv, one_smul]
+
+lemma stabilizer_orbit_iff {x : α} {g g' : G} :
+  g' ∈ stabilizer G x ↔ (g * g' * g⁻¹ ∈ stabilizer G (g • x)) :=
+show _ • _ = _ ↔ _ • _ = _,
+begin
+  simp only [mul_smul,smul_inv_left],
+  symmetry,
+  exact (bijective g).1.eq_iff
+end
+
+lemma stabilizer_orbit_inv_iff {x : α} {g g' : G} :
+  g' ∈ stabilizer G (g • x) ↔ (g⁻¹ * g' * g ∈ stabilizer G x) :=
+begin
+  symmetry,
+  convert stabilizer_orbit_iff; try { apply_instance },
+  rw [mul_assoc,mul_inv_cancel_right,mul_inv_cancel_left]
+end
+
+def equiv.stabilizer (g : G) (x : α) : stabilizer G x ≃ stabilizer G (g • x) :=
+begin
+  use assume g', ⟨g * g'.1 * g⁻¹, stabilizer_orbit_iff.1 g'.2⟩,         -- to_fun
+  use assume g', ⟨g⁻¹ * g'.1 * g, stabilizer_orbit_inv_iff.1 g'.2⟩,     -- inv_fun
+  all_goals { assume g',
+              apply subtype.eq',
+              simp,
+              rw mul_assoc },
+    { rw inv_mul_cancel_right,
+      exact inv_mul_cancel_left _ _ },
+    { rw mul_inv_cancel_right,
+      exact mul_inv_cancel_left _ _ }
+end
+
+variables G α
+def orbits : Type v := quotient $ orbit_rel G α
+
+local attribute [instance] mul_action.orbit_rel
+local attribute [reducible] lattice.fixed_points
+
+/-- A trivial equivalence used in the proof of Burnside Lemma -/
+def burnside_aux₁ :
+  (Σ g : G, lattice.fixed_points ((•) g : α → α)) ≃
+    (Σ (o : orbits G α) (x : {x // ⟦x⟧ = o}), stabilizer G x.val) :=
+{ to_fun  := λ y, ⟨⟦y.snd.val⟧, ⟨⟨y.snd.val, rfl⟩, ⟨y.fst, y.snd.property⟩⟩⟩,
+  inv_fun := λ y, ⟨y.snd.snd.val, ⟨y.snd.fst.val, y.snd.snd.property⟩⟩,
+  left_inv := λ ⟨g, ⟨x, h⟩⟩, rfl,
+  right_inv := λ ⟨o, ⟨⟨x, hx⟩, ⟨g, h⟩⟩⟩, by subst o
+}
+
+noncomputable theory
+open classical
+
+variables {G α}
+
+/-- Choose an element of the group that sends `o.out` to a specific point of the orbit. -/
+def orbit_choice {o : orbits G α} (x : {x // ⟦x⟧ = o}) : G :=
+  some (quotient.exact $ x.2.trans o.out_eq.symm)
+
+def orbit_choice_spec {o : orbits G α} (x : {x // ⟦x⟧ = o}) :
+  orbit_choice x • o.out = x.val :=
+some_spec (quotient.exact $ x.2.trans o.out_eq.symm)
+
```

Comment:
I would suggest that you give versions of the results below that depend on a variable that specifies an orbit choice.  Then you can have versions that use the choice given by `some (quotient.exact ...)`.  That will allow for computation in some concrete cases.

Reply:
1. Why would someone want to compute this? The particular bijection has no meaning at all.
2. The proof relies on a noncomputable group_equiv_quotient_times_subgroup, and I don't want to rewrite its proof.